src/main/java/htsjdk/samtools/cram/build/Sam2CramRecordFactory.java

```diff
@@ -291,57 +258,52 @@ private void addInsertion(final List<ReadFeature> features, final int zeroBasedP
+     * <li>emit a {@link Substitution} read feature for each ACTGN-ACTGN mismatch.</li>
+     * <li>emit {@link ReadBase} for a non-ACTGN mismatch. The side effect is the quality score stored twice.</li>
+     * <p>
+     * The rest of the method handles quality score read features added if refSNPs is specified.
+     * <p>
+     * IMPORTANT: reference and read bases are always compared for match/mismatch in upper case due to BAM limitations.
+     *
+     * @param alignmentStart       CRAM record alignment start
+     * @param features             a list of read features to add to
+     * @param fromPosInRead        a zero based position in the read to start with
+     * @param alignmentStartOffset offset into the reference array
+     * @param nofReadBases         how many read bases to process
+     * @param bases                the read bases array
+     * @param qualityScore         the quality score array
+     */
+    void addMismatchReadFeatures(final int alignmentStart, final List<ReadFeature> features, final int fromPosInRead, final int
             alignmentStartOffset, final int nofReadBases, final byte[] bases, final byte[] qualityScore) {
-        int oneBasedPositionInRead;
-        final boolean noQS = (qualityScore.length == 0);
+        int oneBasedPositionInRead = fromPosInRead + 1;
+        int refIndex = alignmentStart + alignmentStartOffset - 1;
 
-        int i;
-        boolean qualityAdded;
         byte refBase;
-        for (i = 0; i < nofReadBases; i++) {
-            oneBasedPositionInRead = i + fromPosInRead + 1;
-            final int referenceCoordinates = cramRecord.alignmentStart + i + alignmentStartOffset - 1;
-            qualityAdded = false;
-            if (referenceCoordinates >= refBases.length) refBase = 'N';
-            else refBase = refBases[referenceCoordinates];
-            refBase = Utils.normalizeBase(refBase);
-
-            if (bases[i + fromPosInRead] != refBase) {
-                final Substitution substitution = new Substitution();
-                substitution.setPosition(oneBasedPositionInRead);
-                substitution.setBase(bases[i + fromPosInRead]);
-                substitution.setReferenceBase(refBase);
-
-                features.add(substitution);
-
-                if (noQS) continue;
-            }
-
-            if (noQS) continue;
-
-            if (refSNPs != null) {
-                final byte snpOrNot = refSNPs[referenceCoordinates];
-                if (snpOrNot != 0) {
-                    final byte score = (byte) (QS_asciiOffset + qualityScore[i + fromPosInRead]);
-                    features.add(new BaseQualityScore(oneBasedPositionInRead, score));
-                    qualityAdded = true;
-                    landedRefMaskScores++;
+        for (int i = 0; i < nofReadBases; i++, oneBasedPositionInRead++, refIndex++) {
+            if (refIndex >= refBases.length) refBase = 'N';
```

Comment:
Not sure if legitimate but it happens. I believe c1#bounds.sam from the compliance tests covers this.